src/variation.h

```diff
@@ -161,29 +176,63 @@ Program<T> cross(const Program<T>& root, const Program<T>& other)
                                         );
+
         auto child_ret_type = child_spot.node->data.ret_type;
-        // fmt::print("child_spot : {}\n",child_spot.node->data);
-        // fmt::print("child_ret_type: {}\n",child_ret_type);
-        // pick a subtree to insert
-        // need to pick a node that has a matching output type to the child_spot
+
+        auto allowed_size  = PARAMS["max_size"].get<int>() -
+                             ( child.Tree.size() - child.Tree.size(child_spot) );
+        auto allowed_depth = PARAMS["max_depth"].get<int>() - 
+                             ( child.Tree.max_depth() - child.Tree.depth(child_spot) );
+
+        // fmt::print("child_spot      : {}\n",child_spot.node->data);
+        // fmt::print("child_ret_type  : {}\n",child_ret_type);
+        // fmt::print("child size      : {}\n", child.Tree.size());
+        // fmt::print("child depth     : {}\n", child.Tree.max_depth());
+        // fmt::print("child spot size : {}\n", child.Tree.size(child_spot));
+        // fmt::print("child spot depth: {}\n", child.Tree.max_depth(child_spot));
+        // fmt::print("child spot model: {}\n", child_spot.node->get_model());
+
+        // pick a subtree to insert. Selection is based on other_weights
         vector<float> other_weights(other.Tree.size());
+
+        // iterator to get the size of subtrees inside transform
+        auto other_iter = other.Tree.begin();
+
+        // lambda function to check feasibility of solution and increment the iterator 
+        const auto check_and_incrm = [other, &other_iter, allowed_size, allowed_depth]() -> bool {
+            int s = other.Tree.size(other_iter);
+            int d = other.Tree.depth(other_iter);
+
+            std::advance(other_iter, 1);
+            return (s <= allowed_size) && (d <= allowed_depth);
+        };
+
         std::transform(other.Tree.begin(), other.Tree.end(), 
             other_weights.begin(),
-            [child_ret_type](const auto& n){ 
-                if (n.ret_type == child_ret_type)
+            [child_ret_type, check_and_incrm](const auto& n){
+                // need to pick a node that has a matching output type to the child_spot.
+                // also need to check if swaping this node wouldn't exceed max_size
+                if (check_and_incrm() && (n.ret_type == child_ret_type))
                     return n.get_prob_change(); 
                 else
+                    // setting the weight to zero to indicate a non-feasible crossover point
                     return float(0.0);
                 }
             );
+
         for (const auto& w: other_weights)
         {
             matching_spots_found = w > 0.0;
+
             if (matching_spots_found) 
                 break;
         }
+
         if (matching_spots_found) 
         {
+            // The probability of picking the i-th element is w_i/S, with S being
+            // the sum of all weights. select_randomly works with a weight vector
```

Comment:
the comment about select_randomly should be moved to that doc's function. 

thanks for adding these by the way!

Reply:
I wrote that while I was trying to figure out if setting weights to zero would break the code. This was bugging me because I was afraid that setting some probability to zero (in mutation or crossover) would create strange behavior (or inaccurate probabilities)! It turns out that we can have any value, and values are sampled after normalizing the probabilities. 

I moved it as a comment to the `select_randomly`, as it currently has no doxygen commands.

@lacava I am thinking of writing a `TODO` comment telling to add a simple documentation to that file. Then, during the next week, I want to work on fixing my own TODOs, as well as some simple ones that I found along the code. Also, this is a good opportunity for me to write some documentation and fix my doxygen not building the docs :sweat_smile: